File: problems/树/112.路径总和.py
#!/usr/bin/env python3
"""
112. 路径总和
"""
from typing import Optional
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def has_path_sum(root: Optional[TreeNode], target_sum: int) -> bool:
    if not root: return False
    if not root.left and not root.right: return root.val == target_sum
    return has_path_sum(root.left, target_sum - root.val) or            has_path_sum(root.right, target_sum - root.val)

if __name__ == "__main__":
    root1 = TreeNode(5, TreeNode(4, TreeNode(11, TreeNode(7), TreeNode(2)), None),
                    TreeNode(8, TreeNode(13), TreeNode(4, None, TreeNode(1))))
    print(has_path_sum(root1, 22))  # True
    print(has_path_sum(root1, 5))   # False
    print(has_path_sum(TreeNode(1), 1))  # True
    print(has_path_sum(TreeNode(1), 0))  # False
    print(has_path_sum(None, 0))  # False

File: problems/树/112.路径总和.cpp
#include <iostream>
using namespace std;
struct TreeNode {
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};

bool hasPathSum(TreeNode* root, int targetSum) {
    if (!root) return false;
    if (!root->left && !root->right) return root->val == targetSum;
    return hasPathSum(root->left, targetSum - root->val) ||
           hasPathSum(root->right, targetSum - root->val);
}
int main() {
    TreeNode* root = new TreeNode(5);
    root->left  = new TreeNode(4);
    root->right = new TreeNode(8);
    root->left->left  = new TreeNode(11);
    root->right->left  = new TreeNode(13);
    root->right->right = new TreeNode(4);
    root->left->left->left  = new TreeNode(7);
    root->left->left->right = new TreeNode(2);
    root->right->right->right = new TreeNode(1);
    cout << (hasPathSum(root, 22) ? "True" : "False") << endl;  // True
    cout << (hasPathSum(root, 5) ? "True" : "False") << endl;   // False
    cout << (hasPathSum(new TreeNode(1), 1) ? "True" : "False") << endl;  // True
    cout << (hasPathSum(new TreeNode(1), 0) ? "True" : "False") << endl;  // False
    cout << (hasPathSum(nullptr, 0) ? "True" : "False") << endl;  // False
    return 0;
}
